File: android/app/src/main/java/ai/spatialwalk/scenes/VideoPanel.kt
package ai.spatialwalk.scenes

import ai.spatialwalk.scenes.rtc.AvatarRtcSession
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.foundation.clickable
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/** Gap between the two video tiles. */
private val TileGap = 12.dp

/**
 * Right-hand video area: avatar on top, the student below, both square.
 *
 * The side length is derived from the **available height**: two squares plus the gap
 * between them must fit, hence side = (available height - gap) / 2, then capped by
 * [maxTileSize] so it doesn't get oversized on large screens. Deriving the side from a
 * fixed width instead would overflow and get clipped on tall-aspect devices.
 *
 * Placeholder content lives in [VideoPlaceholder], so swapping in real video only touches
 * that one place.
 */
@Composable
fun VideoPanel(
    modifier: Modifier = Modifier,
    maxTileSize: Dp = 360.dp,
    cameraGranted: Boolean = false,
    micEnabled: Boolean = false,
    session: AvatarRtcSession,
    /** Side length of one square tile, computed once by the caller per screen orientation. */
    tileSize: Dp,
    /** Whether the two tiles sit side by side (side by side in portrait, stacked in landscape). */
    horizontal: Boolean = false,
    /** Whether free talk has started. */
    freeTalk: Boolean = false,
    /** The record-audio permission was denied: the mic goes grey and tapping asks again. */
    micDenied: Boolean = false,
    onToggleFreeTalk: () -> Unit = {},
    /** Whether the student has their own video on. */
    cameraOn: Boolean = true,
    onToggleCamera: () -> Unit = {},
    onConnectedChange: (Boolean) -> Unit = {},
) {
    var avatarReady by remember { mutableStateOf(false) }
    /** Waiting-overlay copy: a fixed message by default, stage progress while connecting,
     * the reason on failure. */
    var overlayText by remember { mutableStateOf(Localization.t.teacherComing) }
    var overlayFailed by remember { mutableStateOf(false) }

    Box(modifier = modifier) {
        val tiles: @Composable () -> Unit = {
            // Top: the avatar. The classroom background sits under the render view since
            // the model has a transparent background.
            VideoTile(
                size = tileSize,
                gradient = listOf(Color(0xFF3D5AFE), Color(0xFF7C4DFF)),
            ) {
                Image(
                    painter = painterResource(R.drawable.classroom_background),
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.size(tileSize),
                )
                AvatarStage(
                    modifier = Modifier.size(tileSize),
                    micEnabled = micEnabled,
                    session = session,
                    onConnectedChange = onConnectedChange,
                    // Stage progress is logged, not shown: the waiting overlay keeps one
                    // fixed message rather than exposing internal progress like
                    // "downloading 20%" to the student.
                    onStatus = {},
                    onReadyChange = { avatarReady = it },
                    onFailure = {
                        overlayText = it
                        overlayFailed = true
                    },
                )
                // Cover with an overlay while the model loads and connects, so there is
                // no long blank stretch.
                if (!avatarReady) {
                    TeacherComingOverlay(tileSize, overlayText, spinning = !overlayFailed)
                }
                // Mic toggle overlaid on the bottom right of the teacher's tile: red means
                // the mic is off, green means the call is live.
                MicButton(
                    active = freeTalk && micEnabled,
                    denied = micDenied,
                    onClick = onToggleFreeTalk,
                    modifier = Modifier
                        .align(Alignment.BottomEnd)
                        .padding(10.dp),
                )
            }
            // Bottom: the student. Keep the placeholder when the camera permission is
            // missing or the student turned their own video off.
            VideoTile(
                size = tileSize,
                gradient = listOf(Color(0xFF00897B), Color(0xFF26A69A)),
            ) {
                if (cameraGranted && cameraOn) {
                    CameraPreview(modifier = Modifier.size(tileSize))
                } else {
                    VideoPlaceholder(tileSize, Localization.t.me)
                }
                // Camera toggle on the bottom right of the student's tile, mirroring the
                // mic on the teacher's. Hidden without the permission — there is no video
                // to turn off, so the button would do nothing.
                if (cameraGranted) {
                    CameraButton(
                        active = cameraOn,
                        onClick = onToggleCamera,
                        modifier = Modifier
                            .align(Alignment.BottomEnd)
                            .padding(10.dp),
                    )
                }
            }
        }

        if (horizontal) {
            Row(horizontalArrangement = Arrangement.spacedBy(TileGap)) { tiles() }
        } else {
            Column(verticalArrangement = Arrangement.spacedBy(TileGap)) { tiles() }
        }
    }
}

/**
 * Mic toggle, overlaid on the bottom right of the teacher's tile.
 *
 * Red = mic off (tap to enter free talk and open the mic), green = call live (tap to hang
 * up). When the permission was denied it shows a grey crossed-out mic, and tapping asks
 * for the permission again.
 */
@Composable
private fun MicButton(
    active: Boolean,
    denied: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val background = when {
        denied -> Color(0xFF9E9E9E)
        active -> Color(0xFF2E9E5B)
        else -> Color(0xFFD64545)
    }
    Box(
        modifier = modifier
            .size(44.dp)
            .clip(CircleShape)
            .background(background)
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        // Emoji rather than material-icons — pulling in the whole icon library for a
        // single glyph isn't worth it.
        Text(
            text = if (denied) "🚫" else "🎙",
            fontSize = 20.sp,
        )
    }
}

/**
 * Camera toggle, overlaid on the bottom right of the student's tile.
 *
 * Same color semantics as [MicButton]: green = video on, red = off. It only affects the
 * local preview and involves no RTC publishing — the student's video is never sent
 * upstream in the first place.
 */
@Composable
private fun CameraButton(
    active: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .size(44.dp)
            .clip(CircleShape)
            .background(if (active) Color(0xFF2E9E5B) else Color(0xFFD64545))
            .clickable(onClick = onClick),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = if (active) "📹" else "🚫",
            fontSize = 20.sp,
        )
    }
}

/**
 * Waiting overlay shown until the avatar is in place.
 *
 * Normally it spins and shows the current stage (waiting for network / downloading /
 * connecting); on failure it stops spinning and shows the reason, so a fault doesn't just
 * look like a spinner that never ends.
 */
@Composable
private fun TeacherComingOverlay(size: Dp, text: String, spinning: Boolean) {
    Column(
        modifier = Modifier
            .size(size)
            .background(Color.Black.copy(alpha = 0.45f)),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(10.dp, Alignment.CenterVertically),
    ) {
        if (spinning) {
            CircularProgressIndicator(
                color = Color.White,
                strokeWidth = 2.dp,
                modifier = Modifier.size(size * 0.12f),
            )
        }
        Text(
            text = text,
            color = Color.White,
            fontSize = (size.value * 0.055f).sp,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

/** One square video tile; the side length comes from the caller and [content] fills it. */
@Composable
private fun VideoTile(
    size: Dp,
    gradient: List<Color>,
    content: @Composable () -> Unit,
) {
    Box(
        modifier = Modifier
            .size(size)
            .clip(RoundedCornerShape(16.dp))
            .background(Brush.linearGradient(gradient)),
    ) {
        content()
    }
}

/** Placeholder for video content; replace this when wiring in a real feed. The avatar
 * circle scales proportionally with the tile's side length. */
@Composable
private fun VideoPlaceholder(tileSize: Dp, avatarText: String) {
    Box(
        modifier = Modifier.size(tileSize),
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .size(tileSize * 0.32f)
                .clip(CircleShape)
                .background(Color.White.copy(alpha = 0.22f)),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = avatarText,
                color = Color.White,
                fontSize = (tileSize.value * 0.12f).sp,
                fontWeight = FontWeight.Bold,
            )
        }
    }
}

File: ios/SpatiusScenes/VideoPanel.swift
import SwiftUI

/// Gap between the two video tiles.
private let tileGap: CGFloat = 12

/// The avatar and student tiles, both square. In landscape they stack vertically on the
/// right; in portrait they sit in a row across the top.
///
/// The side length is computed once by `ClassroomView` from the screen orientation and
/// passed in — this view does not derive it itself, since computing it in two places
/// invites the two from disagreeing.
struct VideoPanel: View {
    @ObservedObject var session: AvatarRtcSession
    /// How the two tiles are arranged: a column on the right in landscape, a row at the
    /// top in portrait.
    var axis: Axis = .vertical
    /// Side length of one square tile, computed once by the parent from the orientation.
    var tileSize: CGFloat
    /// Whether free talk has started.
    var freeTalk: Bool = false
    /// Whether the recording permission has been granted.
    var micGranted: Bool = false
    /// Permission was denied: the mic goes grey, and tapping prompts for it again.
    var micDenied: Bool = false
    var onToggleFreeTalk: () -> Void = {}
    /// Whether the student's own camera is on.
    var cameraOn: Bool = true
    var onToggleCamera: () -> Void = {}

    var body: some View {
        // The parent computes the side length once and we set the size directly, rather
        // than negotiating through aspectRatio / fixedSize: a scaledToFill image inside
        // the ZStack stretches the container to the image's ratio, and an aspectRatio
        // on the outside then gets dragged along by the content, turning the square
        // into a rectangle.
        Group {
            if axis == .vertical {
                VStack(spacing: tileGap) { tiles }
            } else {
                HStack(spacing: tileGap) { tiles }
            }
        }
    }

    @ViewBuilder
    private var tiles: some View {
            // Top: the avatar. The classroom background sits under the render view,
            // since the model has a transparent background.
            //
            // Two layers: crop the picture to a square first, then overlay the mic.
            //
            // The mic cannot go in the same ZStack: scaledToFill on the background
            // stretches the ZStack's intrinsic size, and .frame only constrains the
            // displayed box — the inner alignment is still computed against the
            // stretched area, so the bottom-right corner gets pushed outside the box
            // and clipped. That is why only a sliver of it used to show on the left.
            ZStack(alignment: .bottomTrailing) {
                ZStack {
                    Image("classroom_background")
                        .resizable()
                        .scaledToFill()
                    AvatarStage(session: session)
                    if !session.isReady {
                        TeacherComingOverlay(detail: session.status, failed: session.hasFailed)
                    }
                }
                .frame(width: tileSize, height: tileSize)
                .clipShape(RoundedRectangle(cornerRadius: 16))

                MicButton(
                    active: freeTalk && micGranted,
                    denied: micDenied,
                    onTap: onToggleFreeTalk
                )
                .padding(10)
            }
            .frame(width: tileSize, height: tileSize)

            // Bottom: the student. Two layers as well, for the same reason — the button
            // goes outside the clip, otherwise it is pushed out of the box (the inner
            // content stretches the ZStack's intrinsic size).
            ZStack(alignment: .bottomTrailing) {
                Group {
                    if cameraOn {
                        CameraPreview()
                    } else {
                        // With the camera off, leave a solid tile and show no preview.
                        Color.clear
                    }
                }
                .frame(width: tileSize, height: tileSize)
                .background(
                    LinearGradient(
                        colors: [Color(red: 0, green: 0.54, blue: 0.48),
                                 Color(red: 0.15, green: 0.65, blue: 0.60)],
                        startPoint: .topLeading, endPoint: .bottomTrailing
                    )
                )
                .clipShape(RoundedRectangle(cornerRadius: 16))

                CameraButton(active: cameraOn, onTap: onToggleCamera)
                    .padding(10)
            }
            .frame(width: tileSize, height: tileSize)
    }
}

/// The mic toggle, overlaid on the bottom-right of the teacher tile.
///
/// Red = mic off (tap to enter free talk and go live), green = in conversation (tap to
/// hang up). Grey when permission was denied; tapping prompts for it again.
private struct MicButton: View {
    let active: Bool
    let denied: Bool
    let onTap: () -> Void

    private var background: Color {
        if denied { return Color(white: 0.62) }
        return active
            ? Color(red: 0.18, green: 0.62, blue: 0.36)
            : Color(red: 0.84, green: 0.27, blue: 0.27)
    }

    var body: some View {
        Button(action: onTap) {
            Text(denied ? "🚫" : "🎙")
                .font(.system(size: 20))
                .frame(width: 44, height: 44)
                .background(background)
                .clipShape(Circle())
        }
        .buttonStyle(.plain)
    }
}

/// The camera toggle, overlaid on the bottom-right of the student tile.
///
/// Same color semantics as `MicButton`: green = camera on, red = off. It only affects
/// the local preview and involves no RTC publishing — the student's video is never sent
/// upstream in the first place.
private struct CameraButton: View {
    let active: Bool
    let onTap: () -> Void

    var body: some View {
        Button(action: onTap) {
            Text(active ? "📹" : "🚫")
                .font(.system(size: 20))
                .frame(width: 44, height: 44)
                .background(active
                    ? Color(red: 0.18, green: 0.62, blue: 0.36)
                    : Color(red: 0.84, green: 0.27, blue: 0.27))
                .clipShape(Circle())
        }
        .buttonStyle(.plain)
    }
}

/// The waiting overlay shown before the avatar is in place.
///
/// Downloading the model plus connecting RTC takes several seconds, during which the
/// tile is empty and needs to say something. It shows a single line and does not expose
/// internal stages such as download progress.
private struct TeacherComingOverlay: View {
    /// The stage message. Hidden during the normal flow; on failure the reason is shown
    /// to make troubleshooting easier.
    let detail: String
    let failed: Bool

    private var failureText: String? { failed ? detail : nil }

    var body: some View {
        VStack(spacing: 10) {
            ProgressView().tint(.white)
            Text(Localization.shared.t.teacherComing)
                .font(.system(size: 15, weight: .semibold))
                .foregroundStyle(.white)
            if let failureText {
                Text(failureText)
                    .font(.system(size: 12))
                    .foregroundStyle(.white.opacity(0.85))
                    .multilineTextAlignment(.center)
                    .padding(.horizontal, 12)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(Color.black.opacity(0.45))
    }
}
